File: Algorithms/p173_Binary_Search_Tree_Iterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.my_stack = []
        while root:
            self.my_stack.append(root)
            root = root.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return bool(self.my_stack)

    def next(self):
        """
        :rtype: int
        """
        val = None
        if self.my_stack:
            p = self.my_stack.pop()
            val = p.val
            p = p.right
            while p:
                self.my_stack.append(p)
                p = p.left
        return val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())

File: Algorithms/p173_Binary_Search_Tree_Iterator/p173_Binary_Search_Tree_Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    stack<TreeNode*> my_stack;
    
    BSTIterator(TreeNode *root) {
        while(root){
            my_stack.push(root);
            root = root->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !my_stack.empty();
    }

    /** @return the next smallest number */
    int next() {
        int val;
        if(!my_stack.empty()){
            TreeNode* p = my_stack.top();
            my_stack.pop();
            val = p->val;
            p = p->right;
            while(p){
                my_stack.push(p);
                p = p->left;
            }
        }
        return val;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */
